### ISSUE TITLE
Select distinct not working for expressions

### ISSUE THREAD

**Issue by stephenrauch** [web2py/pydal#256]
__The book says__:

Notice that distinct can also be an expression for example:

```
>>> for row in db().select(db.person.name,distinct=db.person.name):
        print row.name
Alex
Bob
Carl
```

I take that to mean that this should work:

``` python
db.define_table('tt', Field('aa'), Field('bb', default='bb'))
db.tt.insert(aa='1111')
db.tt.insert(aa='1112')
db.tt.insert(aa='1112')
result = db().select(db.tt.aa, distinct=True)
result = db().select(db.tt.aa, distinct=db.tt.aa)
```

The first select finishes without error, however the second select has problems.  On sqlite I get the following:

```
  OperationalError: near "ON": syntax error
```

On mssql I get the similar:

```
  ProgrammingError: ('42000', "[42000] [Microsoft][SQL Server Native Client 11.0][SQL Server]Incorrect syntax near the keyword 'ON'. (156) (SQLExecDirectW)")
```

The generated sql is:

```
SELECT DISTINCT ON (tt.aa) tt.aa FROM tt;
```

I am trying to verify that recent changes to the MongoDB driver will support this behavior, and was trying to generate test cases that would work with mssql and sqlite. I have failed to do so.  Any help would be appreciated.

**Comment by gi0baro** [comment]
Maybe @ilvalle and @niphlod can help on this.

**Comment by ilvalle** [comment]
Given this:

```

db.define_table('tt', Field('aa'), Field('bb', default='bb'))

db.tt.insert(aa='1111', bb='a')

db.tt.insert(aa='1112', bb='b')

db.tt.insert(aa='1112', bb='c')

db().select(db.tt.aa, db.tt.bb, distinct=db.tt.aa)

```

Which db.tt.bb will be coupled to the 'distinct' one of db.tt.aa ? It basically depends on the backend used.


```distinct=db.tt.aa``` works on postgres, which has some limit in selecting fields in a group by clause. For example on postgres the following works:

```

db().select(db.tt.aa, db.tt.bb, distinct=db.tt.aa)

```

while this is not allowed:

```

db().select(db.tt.aa, db.tt.bb, distinct=db.tt.aa)

```

On other backends (es: sqlite) both queries return something.


Given these premises, my advice is that if you want to run distinct on a specific field you should use groupby. For what concern mongodb, I would go with the basic approach ```distinct=db.tt.aa```

**Comment by stephenrauch** [comment]
@ilvalle 

Thanks for the comment. 

I am aware the a distinct() call that does not resolve to unique results will give differing results in different situations or backends. I am also aware that groupby is generally better to use. In fact distinct uses the '$group' pipeline stage on Mongo. But that is not my question. My question is: does it even work? Because on both of the SQL backends I have available (MSSQL and SQLite) it does not work. Any combination of your example I try on sqlite bombs with the error:

```
OperationalError: near ".": syntax error
```

In addition, You gave two examples:

> For example on postgres the following works:

```
db().select(db.tt.aa, db.tt.bb, distinct=db.tt.aa)
```

> while this is not allowed:

```
db().select(db.tt.aa, db.tt.bb, distinct=db.tt.aa)
```

These two examples appear to me to be identical. So I am confused.

To date I have tried to validate all of the functionality I have added to the mongo driver with test cases. I have also tried to insure that all of the test cases also work on sqlite and mssql. But in this case I failed to get anything other than distinct=True to work on either driver.

I am far from an expert on SQL, and so have been relying on the test suite and the book as the defacto specification for my mongo driver work.

At this point I have already submitted the mongo code and tests.  I am trying to back port all of my added tests which are appropriate to SQL to the sql.py test file, and this test doesn't work.  During the back port I will simply ignore the test for distinct is something other than True.

Thanks.
